src/GafferScene/AttributeQuery.cpp

```diff
@@ -0,0 +1,1197 @@
+		}
+	}
+}
+
+IECore::InternedString g_defPlugName( "default" );
+IECore::InternedString g_valPlugName( "value" );
+
+} // namespace
+
+namespace GafferScene
+{
+
+size_t AttributeQuery::g_firstPlugIndex = 0;
+
+GAFFER_NODE_DEFINE_TYPE( AttributeQuery );
+
+AttributeQuery::AttributeQuery( std::string const& name )
+: Gaffer::ComputeNode( name )
+{
+	storeIndexOfNextChild( g_firstPlugIndex );
+	addChild( new ScenePlug( "scene" ) );
+	addChild( new Gaffer::StringPlug( "location" ) );
+	addChild( new Gaffer::StringPlug( "attribute" ) );
+	addChild( new Gaffer::BoolPlug( "inherit", Gaffer::Plug::In, false ) );
+	addChild( new Gaffer::BoolPlug( "exists", Gaffer::Plug::Out, false ) );
+	addChild( new Gaffer::ObjectPlug( "__internalObject", Gaffer::Plug::Out, IECore::NullObject::defaultNullObject() ) );
+}
+
+AttributeQuery::~AttributeQuery()
+{}
+
+ScenePlug* AttributeQuery::scenePlug()
+{
+	return const_cast< ScenePlug* >(
+		static_cast< AttributeQuery const* >( this )->scenePlug() );
+}
+
+ScenePlug const* AttributeQuery::scenePlug() const
+{
+	return getChild< ScenePlug >( g_firstPlugIndex );
+}
+
+Gaffer::StringPlug* AttributeQuery::locationPlug()
+{
+	return const_cast< Gaffer::StringPlug* >(
+		static_cast< AttributeQuery const* >( this )->locationPlug() );
+}
+
+Gaffer::StringPlug const* AttributeQuery::locationPlug() const
+{
+	return getChild< Gaffer::StringPlug >( g_firstPlugIndex + 1 );
+}
+
+Gaffer::StringPlug* AttributeQuery::attributePlug()
+{
+	return const_cast< Gaffer::StringPlug* >(
+		static_cast< AttributeQuery const* >( this )->attributePlug() );
+}
+
+Gaffer::StringPlug const* AttributeQuery::attributePlug() const
+{
+	return getChild< Gaffer::StringPlug >( g_firstPlugIndex + 2 );
+}
+
+Gaffer::BoolPlug* AttributeQuery::inheritPlug()
+{
+	return const_cast< Gaffer::BoolPlug* >(
+		static_cast< AttributeQuery const* >( this )->inheritPlug() );
+}
+
+Gaffer::BoolPlug const* AttributeQuery::inheritPlug() const
+{
+	return getChild< Gaffer::BoolPlug >( g_firstPlugIndex + 3 );
+}
+
+Gaffer::BoolPlug* AttributeQuery::existsPlug()
+{
+	return const_cast< Gaffer::BoolPlug* >(
+		static_cast< AttributeQuery const* >( this )->existsPlug() );
+}
+
+Gaffer::BoolPlug const* AttributeQuery::existsPlug() const
+{
+	return getChild< Gaffer::BoolPlug >( g_firstPlugIndex + 4 );
+}
+
+Gaffer::ObjectPlug* AttributeQuery::internalObjectPlug()
+{
+	return const_cast< Gaffer::ObjectPlug* >(
+		static_cast< AttributeQuery const* >( this )->internalObjectPlug() );
+}
+
+Gaffer::ObjectPlug const* AttributeQuery::internalObjectPlug() const
+{
+	return getChild< Gaffer::ObjectPlug >( g_firstPlugIndex + 5 );
+}
+
+bool AttributeQuery::isSetup() const
+{
+	return ( defaultPlug() != 0 ) && ( valuePlug() != 0 );
+}
+
+bool AttributeQuery::canSetup( Gaffer::Plug const* const plug ) const
+{
+	bool const success =
+		( plug != 0 ) &&
+		( plug->direction() == Gaffer::Plug::In ) &&
+		( plug->node() != this ) &&
```

Comment:
The node check prevents the user making a connection from the inherit plug of the AttributeQuery node itself to the plus widget which the ui otherwise allows. The ui also allows a user to grag in out plug to the plus widget.

Reply:
It's a slightly academic distinction, but `setup()` is a separate step from subsequently making the connection. `setup()` just needs a template plug of the right type, and won't make any connections to it. It's `PlugAdder::createConnection()` that makes the connections, so `canCreateConnection()` is the logical point of responsibility for these checks. I made some related comments on the PlugAdder code below...